### PR TITLE
Parsec TPM Evidence

### PR DESCRIPTION
**Pull request by yogeshbdeshpande**

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -0,0 +1,20 @@
+module parsectpm
+
+go 1.20
```
Comment (thomas-fossati) [inline go.mod]: are we using any 1.20 feature? If not, 1.19 should do.

File: .github/workflows/ci-go-cover.yml [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+jobs:
+  cover:
+    name: Coverage
+    runs-on: ubuntu-latest
+    env:
+      GO111MODULE: on
+    steps:
+    - uses: actions/setup-go@v3
+      with:
+        go-version: "1.18"
```
Comment (thomas-fossati) [inline .github/workflows/ci-go-cover.yml:28]: This should be aligned with go.mod.

(Same for the other CI files.)

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,55 @@
+.DEFAULT_GOAL := test
```
Comment (thomas-fossati) [inline Makefile:1]: maybe use veraison/ear as a template

File: test/evidence.json [hunk 4]
```diff
@@ -0,0 +1 @@
+{"kat":{"tpmVer":"2.0","kid":"AAAAAAAAAAAAAAAAAAAAAA==","alg":11,"sig":"ABgACwAgYVIg2f6N0J3TUvk++EsBuMe9HkqffGkUIYgnoKMjgF8AIGeHvWZKgjQLlE/He1KVtZi7NRVUtEY+zmI+pJMYOjNo","pubArea":"ACMACwAEAGAAAAAQABgACwADABAAIPlsULbsXs4mMf2dnwyH887yIrZ0pbEsws4scDM1ceZFACABdbDMrJo+kEQUPqu/fjSD32i5nSFd8URWhQA7SI0lyQ==","certInfo":"/1RDR4AXAAQAAABkAAAAAAAAAAAAAwAAAAQAAAAFBgAAAAAAAAAHAAQAAAABAAQAAAAC"},"pat":{"tpmVer":"2.0","kid":"AAAAAAAAAAAAAAAAAAAAAA==","alg":11,"sig":"ABgACwAg6wPjajQKQRKQor8KdMZKmtAFyQKZrRbkD8BfQgI9J+8AINskdj3glCLyhAMlhhwJjl2MzWI5G4P0XvRFFTLkin2n","attestInfo":"/1RDR4AYAAQAAABkABNBQkFCQUJBQkFCQUJBQkFCQUJBAAAAAAAAAAMAAAAEAAAABQYAAAAAAAAABwAAAAEACwMOAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"}}
```
Comment (thomas-fossati) [inline test/evidence.json]: could you indent this please?

**Comment by yogeshbdeshpande** [comment]
Please hold off review as need to make changes as per recent discussions on removing the AlgID in the interface

File: README.md [hunk 5]
```diff
@@ -1,3 +1,15 @@
 # parsec-tpm
 
-This is a Work In Progress Library to Encode and Decode Parsec TPM Attestation Token
+A Library to Encode and Decode Parsec TPM Evidence as
+detailed in [attested-tls-poc](github.com/CCC-Attestation/attested-tls-poc)
```
Comment (thomas-fossati) [inline README.md]:
```suggestion
A Library to handle Parsec TPM Evidence as
detailed in [attested-tls-poc](https://github.com/CCC-Attestation/attested-tls-poc)
```

File: README.md [hunk 6]
```diff
@@ -1,3 +1,15 @@
 # parsec-tpm
 
-This is a Work In Progress Library to Encode and Decode Parsec TPM Attestation Token
+A Library to handle Parsec TPM Evidence as
+detailed in [attested-tls-poc](https://github.com/CCC-Attestation/attested-tls-poc)
+
+It provides following features
+
+* Encode a Parsec TPM Evidence Collection, containing a Key Attestation and Platform Attestation
```
Comment (thomas-fossati) [inline README.md]:
```suggestion
* Encode a Parsec TPM Evidence Collection object, containing a Key Attestation and Platform Attestation 
```

File: README.md [hunk 7]
```diff
@@ -1,3 +1,15 @@
 # parsec-tpm
 
-This is a Work In Progress Library to Encode and Decode Parsec TPM Attestation Token
+A Library to handle Parsec TPM Evidence as
+detailed in [attested-tls-poc](https://github.com/CCC-Attestation/attested-tls-poc)
+
+It provides following features
+
+* Encode a Parsec TPM Evidence Collection, containing a Key Attestation and Platform Attestation 
+information to CBOR and JSON formats
```
Comment (thomas-fossati) [inline README.md]:
```suggestion
evidence and metadata to CBOR and JSON formats
```

File: README.md [hunk 8]
```diff
@@ -1,3 +1,15 @@
...
+
+* Decode a CBOR and JSON formatted Parsec TPM Evidence into a Parsec TPM Evidence Collection
```
Comment (thomas-fossati) [inline README.md]:
```suggestion
* Decode a CBOR and JSON formatted Parsec TPM Evidence into a Parsec TPM Evidence Collection object
```

File: README.md [hunk 9]
```diff
@@ -1,3 +1,15 @@
...
+
+* Verify the signature on the Key and Platform Attestation elements using the supplied public key
+
+* Sign the Supplied Key and/or Platform Attestation elements using the supplied key
```
Comment (thomas-fossati) [inline README.md]:
```suggestion
* Sign the supplied Key and/or Platform Attestation data using the supplied private key
```

File: common.go [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+
+	tpm2 "github.com/google/go-tpm/tpm2"
+	"github.com/google/go-tpm/tpmutil"
+	"github.com/veraison/eat"
+	"github.com/veraison/go-cose"
+	"github.com/veraison/swid"
+)
+
+const (
+	DefaultTPMHandle = tpmutil.Handle(100)
```
Comment (thomas-fossati) [inline common.go:21]: why is this needed?

File: common.go [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+	ad.ClockInfo = tpm2.ClockInfo{
+		Clock:        3,
+		ResetCount:   4,
+		RestartCount: 5,
+		Safe:         6,
+	}
+	ad.FirmwareVersion = 7
+	ad.QualifiedSigner = tpm2.Name{
+		Handle: &signer}
+
```
Comment (thomas-fossati) [inline common.go]: please check if there is a similar function exposed by the TPM library

File: common.go [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+}
+
+func computeHash(alg tpm2.Algorithm, data []byte) ([]byte, error) {
+	h, err := alg.Hash()
+	if err != nil {
+		return nil, fmt.Errorf("unable to compute hash algorthm from algorithm")
```
Comment (thomas-fossati) [inline common.go]:
```suggestion
		return nil, fmt.Errorf("unable to compute hash algorthm from algorithm: %w", err)
```

File: common.go [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+	hh := h.New()
+	if _, err := hh.Write(data); err != nil {
+		return nil, err
+	}
+	return hh.Sum(nil), nil
+}
+
+type Algo int
+
+func mapCoseAlgToTpmdHash(coseAlg cose.Algorithm) tpm2.Algorithm {
```
Comment (thomas-fossati) [inline common.go]:
```suggestion
func mapCoseAlgToTpmHash(coseAlg cose.Algorithm) tpm2.Algorithm {
```

File: common.go [hunk 14]
```diff
@@ -0,0 +1,202 @@
...
+	var alg tpm2.Algorithm
+	switch coseAlg {
+	case cose.AlgorithmES256:
+		alg = tpm2.AlgSHA256
+	case cose.AlgorithmES384:
+		alg = tpm2.AlgSHA384
+	case cose.AlgorithmES512:
+		alg = tpm2.AlgSHA512
+	}
```
Comment (thomas-fossati) [inline common.go:75]: we need to be explicit that:
1. we return 0 if no mapping is found
2. 0 is an invalid tpm2 algorithm (use `UnSupportedAlg`)

File: common.go [hunk 15]
```diff
@@ -0,0 +1,202 @@
...
+	return alg
+
```
Comment (thomas-fossati) [inline common.go]:
```suggestion
```

File: common.go [hunk 16]
```diff
@@ -0,0 +1,202 @@
...
+}
+func swidHashAlgToTPMAlg(algID uint64) tpm2.Algorithm {
```
Comment (thomas-fossati) [inline common.go:79]:
```suggestion

func swidHashAlgToTPMAlg(algID uint64) tpm2.Algorithm {
```

File: common.go (hunk 13 above)
Comment (thomas-fossati) [inline common.go]: we don't need to go through COSE.  We can create a const enumeration of the TPM algorithms we support and expose that type to the user.

File: common.go [hunk 17]
```diff
@@ -0,0 +1,202 @@
...
+	switch algID {
+	case swid.Sha256:
+		return tpm2.AlgSHA256
+	case swid.Sha384:
+		return tpm2.AlgSHA384
+	case swid.Sha512:
+		return tpm2.AlgSHA512
+
```
Comment (thomas-fossati) [inline common.go]:
```suggestion
```

File: common.go [hunk 18]
```diff
@@ -0,0 +1,202 @@
...
+	case swid.Sha3_256:
+		return tpm2.AlgSHA3_256
+	case swid.Sha3_384:
+		return tpm2.AlgSHA3_384
+	case swid.Sha3_512:
+		return tpm2.AlgSHA3_512
+	}
+	return 0
+}
+func tpmHashAlgToSWIDHash(algID tpm2.Algorithm) uint64 {
```
Comment (thomas-fossati) [inline common.go:98]:
```suggestion

func tpmHashAlgToSWIDHash(algID tpm2.Algorithm) uint64 {
```

File: common.go [hunk 19]
```diff
@@ -0,0 +1,202 @@
...
+	switch algID {
+
```
Comment (thomas-fossati) [inline common.go]:
```suggestion
```

File: common.go [hunk 20]
```diff
@@ -0,0 +1,202 @@
...
+	default:
+		return UnSupportedAlg
+	}
+}
+
+func verify(key crypto.PublicKey, data []byte, sig []byte) error {
+
+	s, err := tpm2.DecodeSignature(bytes.NewBuffer(sig))
+	if err != nil {
+		return fmt.Errorf("unable to decode the signature %w", err)
```
Comment (thomas-fossati) [inline common.go]:
```suggestion
		return fmt.Errorf("unable to decode the signature: %w", err)
```

File: common.go [hunk 21]
```diff
@@ -0,0 +1,202 @@
...
+
+		hdata, err := computeHash(ha, data)
+		if err != nil {
+			return fmt.Errorf("unable to compute hash for input data")
+		}
+
+		// Get the Signature bytes for RSA
+		sign := s.RSA.Signature
+
+		return rsa.VerifyPKCS1v15(vk, h, hdata, sign)
```
Comment (thomas-fossati) [inline common.go]: it is not used, drop it, I'll rot quickly

File: common.go [hunk 22]
```diff
@@ -0,0 +1,202 @@
...
+
+	case tpm2.AlgECDSA:
+		ha := s.ECC.HashAlg
+		_, err := ha.Hash()
+		if err != nil {
+			return fmt.Errorf("not a hash algorithm: %w", err)
+		}
```
Comment (thomas-fossati) [inline common.go]: drop it, it's done again later in `computeHash()`

File: common.go [hunk 23]
```diff
@@ -0,0 +1,202 @@
...
+
+		hdata, err := computeHash(ha, data)
+		if err != nil {
+			return fmt.Errorf("unable to compute hash for input data")
```
Comment (thomas-fossati) [inline common.go]:
```suggestion
			return fmt.Errorf("unable to compute hash for input data: %w", err)
```

File: common.go [hunk 24]
```diff
@@ -0,0 +1,202 @@
...
+		}
+		vk, ok := key.(*ecdsa.PublicKey)
+		if !ok {
+			return fmt.Errorf("invalid public key for algorithm: %v", alg)
+		}
```
Comment (thomas-fossati) [inline common.go]: move this before computing the hash

also, to improve the error message, please add the actual key type (`%T`)

File: common.go [hunk 25]
```diff
@@ -0,0 +1,202 @@
...
+
+		verified := ecdsa.Verify(vk, hdata, s.ECC.R, s.ECC.S)
+		if !verified {
+			return fmt.Errorf("Verification failed")
+
```
Comment (thomas-fossati) [inline common.go:140]:
```suggestion
```

File: common.go [hunk 26]
```diff
@@ -0,0 +1,202 @@
...
+	}
+
+	return nil
+}
+
+func signEcdsa(alg cose.Algorithm, key crypto.PrivateKey, data []byte) ([]byte, error) {
+	sk, ok := key.(*ecdsa.PrivateKey)
+	if !ok {
+		return nil, fmt.Errorf("invalid key")
+	}
```
Comment (thomas-fossati) [inline common.go]: this should be probably done in the caller

File: common.go [hunk 27]
```diff
@@ -0,0 +1,202 @@
...
+	ta := mapCoseAlgToTpmdHash(alg)
```
Comment (thomas-fossati) [inline common.go]: check that `ta` is not `UnsupportedAlg`

File: common.go [hunk 28]
```diff
@@ -0,0 +1,202 @@
...
+	hash, err := computeHash(ta, data)
+	if err != nil {
+		return nil, fmt.Errorf("unable to compute hash : %w", err)
```
Comment (thomas-fossati) [inline common.go]:
```suggestion
		return nil, fmt.Errorf("unable to compute hash: %w", err)
```

File: common.go [hunk 29]
```diff
@@ -0,0 +1,202 @@
...
+		Alg: tpm2.AlgECDSA,
+		ECC: &tpm2.SignatureECC{
+			HashAlg: ta,
+			R:       r,
+			S:       s,
+		},
+	}
+	e, err := sig.Encode()
+	if err != nil {
+		return nil, fmt.Errorf("Signature{%+v}.Encode() returned error: %v", s, err)
```
Comment (thomas-fossati) [inline common.go]:
```suggestion
		return nil, fmt.Errorf("signature encoding failed: %v", err)
```

File: evidence.go [hunk 30]
```diff
@@ -0,0 +1,192 @@
...
+
+type Collection struct {
+	Kat *KAT `cbor:"kat" json:"kat"`
+	Pat *PAT `cbor:"pat" json:"pat"`
+}
+
+// Evidence is a wrapper around TpmParsec Attestation Collection
+type Evidence struct {
+	collection *Collection
+}
```
Comment (thomas-fossati) [inline evidence.go]: why we need `Evidence` here?  It doesn't look like it's doing much.  You could just rename `Collection` to `Evidence`?

File: evidence.go [hunk 31]
```diff
@@ -0,0 +1,192 @@
...
+		return fmt.Errorf("nil token supplied")
+	}
+
+	if err := k.Validate(); err != nil {
+		return fmt.Errorf("validation of key attestation token failed: %w", err)
+	}
+
+	if err := p.Validate(); err != nil {
+		return fmt.Errorf("validation of platform attestation token failed: %w", err)
+	}
```
Comment (thomas-fossati) [inline evidence.go]: This code seems to be reused and it's a good candidate for refactoring.

Maybe make it into a `check()` function on the Evidence object that you can call from different places.

File: evidence.go [hunk 32]
```diff
@@ -0,0 +1,192 @@
...
+	if err := e.collection.Kat.Verify(key); err != nil {
+		return fmt.Errorf("failed to verify signature on key attestation token: %w", err)
+	}
+	if err := e.collection.Pat.Verify(key); err != nil {
+		return fmt.Errorf("failed to verify signature on platform attestation token: %w", err)
+	}
+	return nil
+}
+
+// Sign creates a TPMS Signature bytes by signing over the given data by
```
Comment (thomas-fossati) [inline evidence.go]: this comment needs some TLC

File: kat.go [hunk 33]
```diff
@@ -0,0 +1,326 @@
...
+	CertInfo *[]byte `cbor:"certInfo" json:"certInfo"`
+}
+
+func NewKAT() *KAT {
+	return &KAT{}
+}
+
+func (k *KAT) SetTpmVer(v string) error {
+	if v == "" {
+		return fmt.Errorf("empty string specified")
```
Comment (thomas-fossati) [inline kat.go]:
```suggestion
		return fmt.Errorf("empty string supplied")
```

File: kat.go [hunk 34]
```diff
@@ -0,0 +1,326 @@
...
+	if err := data.Validate(); err != nil {
+		return fmt.Errorf("failed to validate UEID: %w", err)
+	}
+
+	k.KID = &v
+	return nil
+}
+
+func (k *KAT) SetSig(s []byte) error {
+	k.Sig = &s
```
Comment (thomas-fossati) [inline kat.go:50]: you want to check that `len(s)` is not 0

File: kat.go [hunk 35]
```diff
@@ -0,0 +1,326 @@
...
+		return fmt.Errorf("TPM Version not set")
+	} else if *k.TpmVer == "" {
+		return fmt.Errorf("Empty TPM Version")
+	}
+
+	if k.KID == nil {
+		return fmt.Errorf("missing key identifier")
+	}
+	if err := validateKID(*k.KID); err != nil {
+		return fmt.Errorf("invalid KID : %w", err)
```
Comment (thomas-fossati) [inline kat.go]:
```suggestion
		return fmt.Errorf("invalid KID: %w", err)
```

File: kat.go [hunk 36]
```diff
@@ -0,0 +1,326 @@
...
+	}
+
+	if k.CertInfo == nil {
+		return fmt.Errorf("no certificate information")
+	}
+	_, err := tpm2.DecodeAttestationData(*k.CertInfo)
+	if err != nil {
+		return fmt.Errorf("failed to decode supplied certification information %w", err)
```
Comment (thomas-fossati) [inline kat.go]:
```suggestion
		return fmt.Errorf("failed to decode certification information: %w", err)
```

File: kat.go [hunk 37]
```diff
@@ -0,0 +1,326 @@
...
+	}
+
+	if k.Sig == nil {
+		return fmt.Errorf("missing signature")
+	}
+	// Check the signature decode results in a success or not?
+	_, err = tpm2.DecodeSignature(bytes.NewBuffer(*k.Sig))
+	if err != nil {
+		return fmt.Errorf("not a valid signature")
```
Comment (thomas-fossati) [inline kat.go]:
```suggestion
		return fmt.Errorf("not a valid signature: %w", err)
```

File: kat.go [hunk 38]
```diff
@@ -0,0 +1,326 @@
...
+	}
+
+	if k.PubArea == nil {
+		return fmt.Errorf("missing public key information")
+	}
+	_, err = tpm2.DecodePublic(*k.PubArea)
+	if err != nil {
+		return fmt.Errorf("unable to decode the Public Area %w", err)
```
Comment (thomas-fossati) [inline kat.go]:
```suggestion
		return fmt.Errorf("unable to decode the Public Area: %w", err)
```

File: kat.go [hunk 39]
```diff
@@ -0,0 +1,326 @@
...
+}
+
+type TpmCertInfo struct {
+	Name          NameInfo
+	QualifiedName NameInfo
+}
+
+type CertInfo struct {
+	Magic       uint32
+	Type        uint16
```
Comment (thomas-fossati) [inline kat.go]: if magic and type (and any other spec-defined MUST) have predefined values, they need to be enforced in this package.  We can't expect the caller to know the spec.

File: kat.go [hunk 40]
```diff
@@ -0,0 +1,326 @@
...
+	Nonce       []byte
+	TpmCertInfo TpmCertInfo
+}
+
+func (k KAT) DecodeCertInfo() (*CertInfo, error) {
+	certInfo := &CertInfo{}
+
+	if k.CertInfo == nil {
+		return nil, fmt.Errorf("no certification information to decode")
```
Comment (thomas-fossati) [inline kat.go]:
```suggestion
		return nil, errors.New("no certification information to decode")
```

File: kat.go [hunk 41]
```diff
@@ -0,0 +1,326 @@
...
+	}
+
+	ad, err := tpm2.DecodeAttestationData(*k.CertInfo)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode supplied attestation information %w", err)
```
Comment (thomas-fossati) [inline kat.go]:
```suggestion
		return nil, fmt.Errorf("failed to decode supplied attestation information: %w", err)
```

File: kat.go [hunk 42]
```diff
@@ -0,0 +1,326 @@
...
+
+	return certInfo, nil
+}
+
+func getNameInfo(name tpm2.Name) (*NameInfo, error) {
+	nameInfo := &NameInfo{}
+
+	if name.Handle != nil {
+		nameInfo.Handle = uint32(*name.Handle)
+	}
```
Comment (thomas-fossati) [inline kat.go:129]: what is name.Handle is nil?

File: kat.go [hunk 43]
```diff
@@ -0,0 +1,326 @@
...
+
+	alg := tpmHashAlgToSWIDHash(name.Digest.Alg)
+	if alg == 0 {
```
Comment (thomas-fossati) [inline kat.go]: instead of 0 use UnsupportedAlg

File: kat.go [hunk 44]
```diff
@@ -0,0 +1,326 @@
...
+
+// DecodePubArea decodes a given public key, from TPMT_PUBLIC structure
+func (k KAT) DecodePubArea() (crypto.PublicKey, error) {
+
+	if k.PubArea == nil {
+		return nil, fmt.Errorf("no public key parameters to decode")
+	}
+	pub, err := tpm2.DecodePublic(*k.PubArea)
+	if err != nil {
+		return nil, fmt.Errorf("unable to decode the Public Area %w", err)
```
Comment (thomas-fossati) [inline kat.go]:
```suggestion
		return nil, fmt.Errorf("unable to decode the Public Area: %w", err)
```

File: kat.go [hunk 45]
```diff
@@ -0,0 +1,326 @@
...
+	}
+	pk, err := pub.Key()
+	if err != nil {
+		return nil, fmt.Errorf("unable to get crypto public key from TPM2 public key %w", err)
```
Comment (thomas-fossati) [inline kat.go]:
```suggestion
		return nil, fmt.Errorf("unable to get crypto public key from TPM2 public key: %w", err)
```

File: kat.go [hunk 46]
```diff
@@ -0,0 +1,326 @@
...
+	if k.CertInfo == nil || len(*k.CertInfo) == 0 {
+		return fmt.Errorf("no payload content to verify")
+	}
+
+	if k.Sig == nil || len(*k.Sig) == 0 {
+		return fmt.Errorf("no signature on the key attestation token")
+	}
+	err := verify(key, *k.CertInfo, *k.Sig)
+	if err != nil {
+		return fmt.Errorf("failed to verify signature %w", err)
```
Comment (thomas-fossati) [inline kat.go]:
```suggestion
		return fmt.Errorf("failed to verify signature: %w", err)
```

File: kat.go [hunk 47]
```diff
@@ -0,0 +1,326 @@
...
+	}
+	return nil
+}
+
+func (k *KAT) EncodePubArea(alg cose.Algorithm, key crypto.PublicKey) error {
```
Comment (thomas-fossati) [inline kat.go]: we shouldn't use cose here

File: kat.go [hunk 48]
```diff
@@ -0,0 +1,326 @@
...
+					Alg:  tpm2.AlgECDSA,
+					Hash: hashAlg,
+				},
+				CurveID: curve,
+				Point:   tpm2.ECPoint{XRaw: ek.X.Bytes(), YRaw: ek.Y.Bytes()},
+			},
+		}
+		pk, err := p.Encode()
+		if err != nil {
+			return fmt.Errorf("unable to encode a public key")
```
Comment (thomas-fossati) [inline kat.go]:
```suggestion
			return fmt.Errorf("unable to encode a public key: %w", err)
```

File: kat.go [hunk 49]
```diff
@@ -0,0 +1,326 @@
...
+				KeyBits:     2048,
+				ExponentRaw: uint32(rk.E),
+				ModulusRaw:  rk.N.Bytes(),
+			},
+		}
+		pk, err := p.Encode()
+		if err != nil {
+			return fmt.Errorf("unable to encode a public key")
+		}
+		k.PubArea = &pk
```
Comment (thomas-fossati) [inline kat.go]: drop this

File: kat.go [hunk 50]
```diff
@@ -0,0 +1,326 @@
...
+			Digest: &tpm2.HashValue{
+				Alg:   cAlg,
+				Value: c.TpmCertInfo.QualifiedName.DigInfo.Digest,
+			},
+		},
+	}
+
+	encoded, err := ad.Encode()
+	if err != nil {
+		return fmt.Errorf("unable to encode certify information")
```
Comment (thomas-fossati) [inline kat.go]:
```suggestion
		return fmt.Errorf("unable to encode certify information: %w", err)
```

File: kat.go [hunk 51]
```diff
@@ -0,0 +1,326 @@
...
+		}
+		k.PubArea = &pk
+	}
+
+	return nil
+}
+
+func (k *KAT) EncodeCertInfo(c CertInfo) error {
+	ad := tpm2.AttestationData{}
+	setTpmAttestDefaults(&ad)
```
Comment (thomas-fossati) [inline kat.go]: this could be a NewTpmAttestDefault() that returns an initialised `ad`

File: kat.go [hunk 52]
```diff
@@ -0,0 +1,326 @@
...
+	ad.Magic = c.Magic
+	ad.Type = tpmutil.Tag(c.Type)
+	ad.ExtraData = c.Nonce
+
+	name := tpmutil.Handle(c.TpmCertInfo.Name.Handle)
+	alg := swidHashAlgToTPMAlg(c.TpmCertInfo.Name.DigInfo.HashAlgID)
+	if alg == 0 {
```
Comment (thomas-fossati) [inline kat.go]: use unsupportedAlg

File: kat.go [hunk 53]
```diff
@@ -0,0 +1,326 @@
...
+		return fmt.Errorf("unable to map algorithm")
```
Comment (thomas-fossati) [inline kat.go]: maybe say which alg we weren't able to map

File: kat.go [hunk 54]
```diff
@@ -0,0 +1,326 @@
...
+	}
+	cName := tpmutil.Handle(c.TpmCertInfo.QualifiedName.Handle)
+	cAlg := swidHashAlgToTPMAlg(c.TpmCertInfo.QualifiedName.DigInfo.HashAlgID)
+	if cAlg == 0 {
```
Comment (thomas-fossati) [inline kat.go]: ditto

File: kat.go [hunk 55]
```diff
@@ -0,0 +1,326 @@
...
+		return fmt.Errorf("unable to map algorithm")
```
Comment (thomas-fossati) [inline kat.go]: ditto

File: common.go (hunk 14 above)
Comment (yogeshbdeshpande) [reply common.go:75]: I forgot to mention that default returns tpm2.AlgUnknown so we are covered here!

File: common.go (hunk 10 above)
Comment (yogeshbdeshpande) [reply common.go:21]: Needed in setting TPM Defaults

File: common.go (hunk 11 above)
Comment (yogeshbdeshpande) [reply common.go]: I have checked, its not a public function to be used!

File: evidence_test.go [hunk 56]
```diff
@@ -0,0 +1,320 @@
...
+func TestEvidence_Verify_RSAKey_Failed(t *testing.T) {
+	e := &Evidence{Kat: mustBuildValidKAT(t), Pat: mustBuildValidPAT(t)}
+	key := generateTestRSAKey(t)
+	pk := key.Public().(*rsa.PublicKey)
+
+	err := e.Verify(pk)
+	expectedErr := "failed to verify signature on key attestation token: failed to verify signature: invalid public key type: *rsa.PublicKey"
+	assert.EqualError(t, err, expectedErr)
+
+}
```
Comment (thomas-fossati) [inline evidence_test.go:267]: left over to be removed

File: test_common.go [hunk 57]
```diff
@@ -0,0 +1,47 @@
...
+	return key
+}
+
+func generateTestRSAKey(t *testing.T) *rsa.PrivateKey {
+	key, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		t.Fatalf("rsa.GenerateKey() error = %v", err)
+	}
+	return key
+}
```
Comment (thomas-fossati) [inline test_common.go:47]: leftover to be removed

**Review by thomas-fossati — APPROVED** [review]
LGTM.  There are a couple of RSA-related leftovers to delete before merging

File: evidence_test.go (hunk 56 above)
Comment (yogeshbdeshpande) [reply evidence_test.go:267]: RSA Key is used to generate negative test scenario

File: test_common.go (hunk 57 above)
Comment (yogeshbdeshpande) [reply test_common.go:47]: Is used in negative test scenario testing

**Pull request merged**